### PR TITLE
Improve Cover HTML page

### PR DESCRIPTION
**Pull request by milmazz**
The idea behind this PR is to improve the current HTML page produced by the `cover` tool, here is an example of the result:

36948266-c2453738-1f9d-11e8-80a4-d542a7dc25a2

### CODE REVIEW

File: lib/tools/src/cover.erl [hunk 1]
```diff
@@ -2525,11 +2523,64 @@ fill1() ->      "..|  ".
...
+   "<td class=\"hits\">",maybe_integer_to_list(N),"</td>\n",
+   "<td class=\"source\"><code>",LineNoNL,"</code></td>\n</tr>\n"].
+
+maybe_integer_to_list(N) when is_integer(N) -> integer_to_list(N);
+maybe_integer_to_list(_) -> "".
+
+read_stylesheet() ->
+    PrivDir = code:priv_dir(?TOOLS_APP),
+    {ok, Css} = file:read_file(filename:join(PrivDir, ?STYLESHEET)),
+    lists:map(fun erlang:binary_to_list/1, binary:split(Css, <<"\n">>, [global])).
```
Comment (josevalim) [inline lib/tools/src/cover.erl]: I believe `cover` is working with iolists, so we don't need to break it apart on multiple lines. Returning `[Css]` should be fine.

File: lib/tools/src/cover.erl [hunk 2]
```diff
@@ -2525,11 +2523,64 @@ fill1() ->      "..|  ".
...
+    "<body>\n"
+    "<h1><code>",OutFile,"</code></h1>\n"].
+
+create_footer(ErlFile, Timestamp) ->
+    ["<footer>File generated from <code>",ErlFile,
+    "</code> by <a href=\"http://erlang.org/doc/man/cover.html\">cover</a> at ",
+    Timestamp,"</footer>\n<table>\n<tbody>\n"].
+
+close_html() ->
+    lists:concat(["</tbody>\n",
```
Comment (josevalim) [inline lib/tools/src/cover.erl]: You probably don't need `lists:concat/1` here. Returning a list of lists should be fine!

Comment (milmazz) [reply lib/tools/src/cover.erl]: @josevalim Thanks for the review!

**Comment by tuncer** [comment]
I welcome the stylistic improvements. Just a few nits:

- Aren't the `line-height` and `padding-left` settings adding avoidable extra whitespace?
- Is it just me or does this reduce contrast?

I don't know if the team will merge the whitespace changes in `cover.erl`, but I suggest to make them separate in order to ease review and git-bisect.

**Comment by milmazz** [comment]
@tuncer I already split the commits as you suggested, now there is one commit that only removes the trailing whitespace, that way would be easy to review this PR.

Regarding your questions:

> Is it just me or does this reduce contrast?

Could you be more specific? are you referring to the sidebar area (left-side)?

> I would use a smaller size for the title. It looks overly large compared to the rest

Yes, I agree with that suggestion. I'll cut the size of the header.

**Comment by josevalim** [comment]
> Could you be more specific? are you referring to the sidebar area (left-side)?

I believe @tuncer is refering to the contrast between green and red lines. The red line in your screenshot is quite dark. It is a matter of style though. Maybe one way to settle the discussion is to pick the green/red colors that GitHub uses on diffs, since most people are used to those anyway.

**Comment by tuncer** [comment]
No, the new green and red were fine. My reaction was due to the screenshot not being sharp. Well, not as sharp as my local font rendering. In fact, I liked the softer red/green and wasn't suggesting to change that. Looking at the screenshot again, I can confirm sufficient contrast. So, I asked if it's just me and answered it myself by realizing the text in the screenshot looks blurrier than what I'm used to on my machines.

To summarize, I'm in favor of non-GitHub soft red and green from the original commit and wouldn't mind if you reverted back to it. And sorry for mistaking the screenshot for low contrast.

Personally, I would also replace the white bg with a softer color (similar to sidebar) to make it easier on the eyes.

This is all my personal take, of course. The OTP team is likely to have different preferences and taste.

**Comment by josevalim** [comment]
I also apologize for my wrong guessing. :) Anyway, now we have two options, so we can wait if the OTP team has any preferences. @milmazz, could you please include a screenshot with your latest changes? Thank you!

**Comment by tuncer** [comment]
The screenshots confirm that I'd prefer the original red/green. GitHub's looks worse to me. Also, I think there's room to shrink the title even further.

Nothing else to comment, if we ignore my personal preference for a non-white BG (similar to sidebar).

**Comment by vassilevsky** [comment]
Hi :)

Indeed, this is related, presentation-wise :) Thanks @tuncer :)

I'd like to find out the rationale behind monospace font choices you made!

```css
code {
  font: 12px "SFMono-Regular", Consolas, "Liberation Mono", Menlo, Courier, monospace;
}

td.source {
  font: 12px monaco, monospace;
}
```

Also, which font is your favourite?

Let's talk fonts!

**Comment by milmazz** [comment]
@vassilevsky I don't have any strong preference here, if I recall correctly I took the same `font-family` that GitHub uses after Jose's suggestion. But I'm open to suggestions.

BTW, I use the Fira Code font in my $EDITOR

**Comment by vassilevsky** [comment]
Isn't Fira Code great? My other absolute favorite is Consolas. Such a well-designed font.

My idea is that users will be much happier when they see their preferred monospace font in Erlang documentation and tools. If they set a nice font in browser settings, but cover will still display something else, that will not be a good experience. And the only way to display their favourite font is to leave only `monospace` as the font family specification. That's what I did in my PR #1824. I think this is the best we can do.

What do you think?

**Comment by vladdu** [comment]
@vassilevsky [slightly unrelated] How do you handle FiraCode's display of "<=" as "less or equal than"?

**Comment by leoliu** [comment]
After a few days of firacode I have to turn off its ligatures to make it usable. I don’t think the font is suitable for erlang and probably will never be.

**Comment by vassilevsky** [comment]
I don't use Fira Code all the time. I *sometimes* set it in the editor, work with it for like a week, and then switch back to Consolas 🙃 because it's just better for me. Oh, and I've just started learning Erlang (reading a lot of docs, so decided to improve them a bit), and I mostly look at Ruby. Not that many symbols there :)

**Comment by tuncer** [comment]
It's hard (if not impossible) to create a typeface with ligatures that work across many languages without being ambiguous or confusing. I couldn't find the spec right now, but I recall that there are newer OpenType features to make ligatures less static and more context specific. You can work around this using different fonts for each language by building variants of Monoid or Iosevka.

When using ligatures, most of the time it's also important that a ligature has the same width/height as the underlying syntax does in the "dumb" editor mode. That is, unless you prefer and are comfortable with variably spaced editing.

It should also be noted that ligatures aren't the only thing you may wish for in your editor. There's also things like superscript/subscript, to give one example. Plus other comfort features one might find useful.

All of this and the spotty ligature and OpenType feature support in editors is why I'd say prettify-symbols (Emacs) and its equivalent in VSCode and friends is the superior and more flexible approach. The flexibility is very useful and being syntax specific means you don't get the wrong symbol. On top of that, it doesn't limit your font choice.

**Comment by tuncer** [comment]
@milmazz, if specifying just `monospace` is the portable way to let the browser select the default fixed width typeface, then I'm all for @vassilevsky's approach. It would additionally avoid a future change. But maybe it doesn't work across all browsers and that's why specific fonts are listed in a certain order.

**Comment by milmazz** [comment]
@tuncer @vassilevsky I forgot to ping you, but some days ago I made the font change that you suggested and now `monospace` is used for code. This is my first PR here, is something more from my end that needs to be done?

**Comment by vassilevsky** [comment]
I've seen your push anyway. Thanks! I'm also new, waiting for maintainers on my PR as well :)

**Comment by tuncer** [comment]
@milmazz, thanks. Assuming you have restored the custom diff colors as found in your first commit, I don't have further comments. Also, to be clear, I'm not part of the OTP team. They will do the "official" review.

**Comment by sirihansen** [comment]
Hello - I'm sorry for the delayed review! I find the discussion above really good and we like the PR, but we have a few comments:

* Please remove the whitespace changes
* The current colors are fine, so don't change back to the original. (Sorry, @tuncer !)
* I find the title is still a bit too big. It really doesn't look so good when the path is long.
* The original code specifies the background color for the body to white. I think this should be specified in the new style sheet as well.

Thanks!

**Comment by tuncer** [comment]
@sirihansen, I'm glad someone else finds the new title too big :). Plus, if we can make the layout more compact, we could fit in more content. I realize this is all a compromise between different screen sizes, but I think one can specify in pure CSS (no JavaScript) to be "responsive". If that's possible here, I can't say.

Regarding diff colors, what I meant is that the initial change that made the colors softer is what I would prefer, not the adjustment to imitate github.com's scheme. So, to be clear, which version do you like? Soft or github?

**Comment by milmazz** [comment]
@sirihansen Thanks for the review, I think I covered all your suggestions, please let me know what do you think:

screen shot 2018-07-10 at 12 49 29 am

**Comment by sirihansen** [comment]
This has run successfully through the nightly tests. Could you please squash all commits to one?

**Comment by milmazz** [comment]
@sirihansen Done.

**Comment by milmazz** [comment]
@sirihansen any feedback after the latest changes?

**Comment by garazdawi** [comment]
During summer Sweden basically shuts down :) Siri will get back to you once the summer holidays are over.

**Comment by sirihansen** [comment]
Sorry for the delay - as Lukas pointed out, I was on vacation. The PR is now merged! Thank you for the contribution!

**Comment by milmazz** [comment]
@garazdawi I didn't know that, good for them :)

@sirihansen I hope you have enjoyed your holidays and don't worry about the delay. Thanks for you review and the approval.
